File: lib/src/main/java/com/sd/lib/vlevel/FVisibleLevel.java
package com.sd.lib.vlevel;

import android.text.TextUtils;
import android.util.Log;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ConcurrentHashMap;

public abstract class FVisibleLevel {
    private static final Map<Class<? extends FVisibleLevel>, FVisibleLevel> MAP_LEVEL = new ConcurrentHashMap<>();

    private static final FVisibleLevelItem EMPTY_ITEM = new FVisibleLevelItem(null, null);
    private final Map<String, FVisibleLevelItem> mMapLevelItem = new ConcurrentHashMap<>();
    private final Map<VisibilityCallback, String> mVisibilityCallbackHolder = new WeakHashMap<>();

    private boolean mIsVisible = true;
    private FVisibleLevelItem mVisibleItem;

    private static boolean sIsDebug;

    protected FVisibleLevel() {
    }

    public static void setDebug(boolean isDebug) {
        sIsDebug = isDebug;
    }

    /**
     * 返回某个等级
     *
     * @param clazz
     * @return
     */
    public static synchronized FVisibleLevel get(Class<? extends FVisibleLevel> clazz) {
        if (clazz == null) {
            throw new NullPointerException("clazz is null");
        }

        if (clazz == FVisibleLevel.class) {
            throw new IllegalArgumentException("clazz is " + clazz.getName());
        }

        FVisibleLevel level = MAP_LEVEL.get(clazz);
        if (level == null) {
            level = createLevel(clazz);
            if (level == null) {
                throw new RuntimeException("create level failed " + clazz.getName());
            }

            MAP_LEVEL.put(clazz, level);

            if (sIsDebug) {
                Log.i(FVisibleLevel.class.getSimpleName(), clazz.getName() + " create +++++");
            }

            level.onCreate();
        }
        return level;
    }

    /**
     * 清空所有等级
     */
    public static synchronized void clearLevel() {
        if (sIsDebug) {
            Log.i(FVisibleLevel.class.getSimpleName(), "clearLevel");
        }

        MAP_LEVEL.clear();
    }

    /**
     * 创建回调
     */
    protected abstract void onCreate();

    /**
     * 创建Item回调
     *
     * @param item
     */
    protected abstract void onCreateItem(FVisibleLevelItem item);

    /**
     * 初始化Item
     *
     * @param items
     */
    public void initItems(String[] items) {
        if (items == null || items.length <= 0) {
            throw new RuntimeException("items is null or empty " + getClass().getName());
        }

        mMapLevelItem.clear();
        for (String item : items) {
            if (TextUtils.isEmpty(item)) {
                throw new RuntimeException("item is empty");
            }

            mMapLevelItem.put(item, EMPTY_ITEM);
        }

        if (sIsDebug) {
            final StringBuilder builder = new StringBuilder();
            builder.append(getClass().getName()).append(" initItems").append("\r\n");
            for (String item : items) {
                builder.append(item).append("\r\n");
            }
            Log.i(FVisibleLevel.class.getSimpleName(), builder.toString());
        }
    }

    /**
     * 返回某个Item
     *
     * @param name
     * @return
     */
    public final FVisibleLevelItem getItem(String name) {
        return getOrCreateItem(name);
    }

    private FVisibleLevelItem getOrCreateItem(String name) {
        if (TextUtils.isEmpty(name)) {
            throw new RuntimeException("name is empty");
        }

        FVisibleLevelItem item = mMapLevelItem.get(name);
        if (item == null) {
            throw new RuntimeException("Item " + name + " was not found in level " + FVisibleLevel.this);
        }

        if (item == EMPTY_ITEM) {
            item = new FVisibleLevelItem(name, FVisibleLevel.this);
            mMapLevelItem.put(name, item);

            if (sIsDebug) {
                Log.i(FVisibleLevel.class.getSimpleName(), getClass().getName() + " create item:" + name);
            }

            onCreateItem(item);
        }
        return item;
    }

    /**
     * 添加回调，弱引用保存回调对象
     *
     * @param callback
     */
    public final void addVisibilityCallback(VisibilityCallback callback) {
        if (callback != null) {
            mVisibilityCallbackHolder.put(callback, "");
        }
    }

    /**
     * 移除回调
     *
     * @param callback
     */
    public final void removeVisibilityCallback(VisibilityCallback callback) {
        if (callback != null) {
            mVisibilityCallbackHolder.remove(callback);
        }
    }

    private Collection<VisibilityCallback> getVisibilityCallbacks() {
        return Collections.unmodifiableCollection(mVisibilityCallbackHolder.keySet());
    }

    /**
     * 等级是否可见
     *
     * @return
     */
    public final boolean isVisible() {
        return mIsVisible;
    }

    /**
     * 返回可见的Item
     *
     * @return
     */
    public final FVisibleLevelItem getVisibleItem() {
        return mVisibleItem;
    }

    /**
     * 设置等级是否可见
     *
     * @param visible
     */
    public final void setVisible(boolean visible) {
        if (mIsVisible != visible) {
            mIsVisible = visible;
            if (sIsDebug) {
                Log.i(FVisibleLevel.class.getSimpleName(), getClass().getName() + " setVisible:" + visible);
            }

            for (VisibilityCallback callback : getVisibilityCallbacks()) {
                callback.onLevelVisibilityChanged(visible, FVisibleLevel.this);
            }
            notifyItemVisibility(visible, mVisibleItem);
        }
    }

    /**
     * 设置Item可见
     *
     * @param name
     */
    public final void visibleItem(String name) {
        if (sIsDebug) {
            Log.i(FVisibleLevel.class.getSimpleName(), getClass().getName() + " visibleItem:" + name + " levelVisible:" + mIsVisible);
        }

        final FVisibleLevelItem item = getOrCreateItem(name);
        final FVisibleLevelItem old = mVisibleItem;
        if (old != item) {
            mVisibleItem = item;

            if (old != null) {
                notifyItemVisibility(false, old);
            }

            if (mIsVisible) {
                notifyItemVisibility(true, item);
            }
        }
    }

    /**
     * 通知可见的Item
     */
    public final void notifyVisibleItem() {
        if (mIsVisible) {
            notifyItemVisibility(true, mVisibleItem);
        }
    }

    private void notifyItemVisibility(boolean visible, FVisibleLevelItem item) {
        if (item == null) {
            return;
        }

        if (sIsDebug) {
            Log.i(FVisibleLevel.class.getSimpleName(), getClass().getName() + " notifyItemVisibility visible:" + visible + " item:" + item.getName());
        }

        if (mMapLevelItem.containsKey(item.getName())) {
            item.notifyVisibility(visible);
        } else {
            throw new RuntimeException("Item " + item.getName() + " was not found in level " + FVisibleLevel.this);
        }
    }

    private static FVisibleLevel createLevel(Class<? extends FVisibleLevel> clazz) {
        try {
            return clazz.newInstance();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        }
        return null;
    }

    public interface VisibilityCallback {
        /**
         * 等级可见状态变化回调
         *
         * @param visible
         * @param level
         */
        void onLevelVisibilityChanged(boolean visible, FVisibleLevel level);
    }
}

File: lib/src/main/java/com/sd/lib/vlevel/FVisibleLevel.kt
package com.sd.lib.vlevel

import android.os.Looper
import android.util.Log
import java.util.*

abstract class FVisibleLevel protected constructor() {
    /** 保存当前等级的Item */
    private val _itemHolder: MutableMap<String, FVisibleLevelItem> = mutableMapOf()

    /** 当前等级是否可见 */
    private var _isVisible = false

    /** 当前等级是否已经被移除 */
    var isRemoved = false
        private set(value) {
            require(value) { "Can not set false to this flag" }
            field = value
        }

    /** 当前Item */
    var currentItem: FVisibleLevelItem = EmptyItem
        private set

    /** 当前等级是否可见 */
    val isVisible: Boolean get() = _isVisible

    /**
     * 等级创建回调
     */
    protected abstract fun onCreate()

    /**
     * Item创建回调
     */
    protected abstract fun onCreateItem(item: FVisibleLevelItem)

    /**
     * 添加Item，跳过重复的Item
     */
    fun addItems(items: Array<String>) {
        checkUiThread()
        if (items.isEmpty()) return
        if (isRemoved) return

        for (item in items) {
            require(item.isNotEmpty()) { "item is empty" }
            if (!_itemHolder.containsKey(item)) {
                _itemHolder[item] = EmptyItem
            }
        }

        logMsg {
            _itemHolder.keys.joinToString(
                prefix = "${this@FVisibleLevel} addItems [",
                separator = ", ",
                postfix = "]",
            )
        }
    }

    /**
     * 获取名称为[name]的Item
     */
    fun getItem(name: String): FVisibleLevelItem {
        return getOrCreateItem(name)
    }

    private fun getOrCreateItem(name: String): FVisibleLevelItem {
        checkUiThread()
        require(name.isNotEmpty()) { "name is empty" }
        if (isRemoved) return EmptyItem

        val cache = requireNotNull(_itemHolder[name]) { "Item ($name) was not found in level ${this@FVisibleLevel}" }
        if (cache != EmptyItem) return cache

        return FVisibleLevelItem(
            name = name,
            level = this@FVisibleLevel,
        ).also { item ->
            _itemHolder[name] = item
            onCreateItem(item)
        }
    }

    /**
     * 设置当前等级是否可见
     */
    fun setVisible(visible: Boolean) {
        checkUiThread()
        if (isRemoved) return
        if (isVisible == visible) return

        _isVisible = visible
        logMsg { "${this@FVisibleLevel} setVisible $visible" }
        notifyItemVisibility(visible, currentItem)
    }

    /**
     * 设置当前等级可见Item为[name]
     */
    fun setCurrentItem(name: String) {
        checkUiThread()
        if (isRemoved) return
        val uuid = if (isDebug) UUID.randomUUID().toString() else ""

        val oldItem = currentItem
        val newItem = getOrCreateItem(name)
        if (newItem == oldItem) return

        currentItem = newItem

        logMsg { "${this@FVisibleLevel} start (${oldItem.name}) -> ($name) currentItem:${currentItem.name} isVisible:$isVisible uuid:$uuid" }
        notifyItemVisibility(false, oldItem)
        notifyItemVisibility(true, newItem)
        logMsg { "${this@FVisibleLevel} finish (${oldItem.name}) -> ($name) currentItem:${currentItem.name} isVisible:$isVisible uuid:$uuid" }
    }

    /**
     * 通知Item的可见状态
     */
    private fun notifyItemVisibility(value: Boolean, item: FVisibleLevelItem) {
        if (isRemoved) return
        if (item == EmptyItem) return

        if (value && !_isVisible) return
        if (value && currentItem != item) return

        if (_itemHolder.containsKey(item.name)) {
            logMsg { "${this@FVisibleLevel} item (${item.name}) -> $value" }
            item.notifyVisibility(value)
        }
    }

    /**
     * 清空Item并设置当前等级为不可见状态
     */
    fun reset() {
        checkUiThread()
        logMsg { "${this@FVisibleLevel} reset" }
        _isVisible = false
        _itemHolder.clear()
        currentItem = EmptyItem
    }

    companion object {
        /** 保存等级对象 */
        private val sLevelHolder: MutableMap<Class<out FVisibleLevel>, FVisibleLevel> = hashMapOf()

        @JvmStatic
        var isDebug = false

        /**
         * 获取等级
         */
        @JvmStatic
        fun get(clazz: Class<out FVisibleLevel>): FVisibleLevel {
            checkUiThread()
            val cache = sLevelHolder[clazz]
            if (cache != null) return cache

            return clazz.newInstance().also { level ->
                sLevelHolder[clazz] = level
                logMsg { "$level +++++" }
                level.onCreate()
            }
        }

        /**
         * 移除等级
         */
        @JvmStatic
        fun remove(clazz: Class<out FVisibleLevel>) {
            checkUiThread()
            sLevelHolder.remove(clazz)?.let { level ->
                logMsg { "$level -----" }
                level.isRemoved = true
            }
        }

        /**
         * 空的Item
         */
        @JvmStatic
        val EmptyItem = FVisibleLevelItem("", object : FVisibleLevel() {
            override fun onCreate() {}
            override fun onCreateItem(item: FVisibleLevelItem) {}
        })
    }
}

internal inline fun logMsg(block: () -> String) {
    if (FVisibleLevel.isDebug) {
        Log.i("FVisibleLevel", block())
    }
}

internal fun checkUiThread() {
    check(Looper.myLooper() == Looper.getMainLooper()) { "You should do this on ui thread." }
}

inline fun <reified T : FVisibleLevel> fVisibleLevel(): FVisibleLevel {
    return FVisibleLevel.get(T::class.java)
}

inline fun <reified T : FVisibleLevel> fVisibleLevelRemove() {
    FVisibleLevel.remove(T::class.java)
}